<a href="https://colab.research.google.com/github/n1az/Intro-to-deep-learning-with-keras/blob/main/Week%205/Final%20Assignment/Peer-graded%20Assignment%3A%20Build%20a%20Regression%20Model%20in%20Keras%20(B).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Download and Clean Dataset

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

We will be using the dataset provided in the assignment

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's read the dataset into a <em>pandas</em> dataframe.

In [2]:
concrete_data = pd.read_csv('sample_data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

#### Let's check how many data points we have.

In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag  ...          Age     Strength
count  1030.000000         1030.000000  ...  1030.000000  1030.000000
mean    281.167864           73.895825  ...    45.662136    35.817961
std     104.506364           86.279342  ...    63.169912    16.705742
min     102.000000            0.000000  ...     1.000000     2.330000
25%     192.375000            0.000000  ...     7.000000    23.710000
50%     272.900000           22.000000  ...    28.000000    34.445000
75%     350.000000          142.950000  ...    56.000000    46.135000
max     540.000000          359.400000  ...   365.000000    82.600000

[8 rows x 9 columns]

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

#### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Coarse Aggregate  Fine Aggregate  Age
0   540.0                 0.0      0.0  ...            1040.0           676.0   28
1   540.0                 0.0      0.0  ...            1055.0           676.0   28
2   332.5               142.5      0.0  ...             932.0           594.0  270
3   332.5               142.5      0.0  ...             932.0           594.0  365
4   198.6               132.4      0.0  ...             978.4           825.5  360

[5 rows x 8 columns]

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

In [10]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

<a id="item1"></a>

<a id="item1"></a>

## Import Keras

#### Let's go ahead and import the Keras library

In [11]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [12]:
from keras.models import Sequential
from keras.layers import Dense

In [13]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [14]:
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [15]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

## Train and Test the Network

Let's call the function now to create our model.

In [16]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs.


In [17]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=2)

Epoch 1/50
23/23 - 14s - loss: 1687.4987
Epoch 2/50
23/23 - 0s - loss: 1670.8235
Epoch 3/50
23/23 - 0s - loss: 1656.0977
Epoch 4/50
23/23 - 0s - loss: 1643.0719
Epoch 5/50
23/23 - 0s - loss: 1631.5956
Epoch 6/50
23/23 - 0s - loss: 1621.4138
Epoch 7/50
23/23 - 0s - loss: 1612.3365
Epoch 8/50
23/23 - 0s - loss: 1604.0355
Epoch 9/50
23/23 - 0s - loss: 1596.5250
Epoch 10/50
23/23 - 0s - loss: 1589.4579
Epoch 11/50
23/23 - 0s - loss: 1582.6826
Epoch 12/50
23/23 - 0s - loss: 1576.0640
Epoch 13/50
23/23 - 0s - loss: 1569.8041
Epoch 14/50
23/23 - 0s - loss: 1563.5409
Epoch 15/50
23/23 - 0s - loss: 1557.2672
Epoch 16/50
23/23 - 0s - loss: 1550.9332
Epoch 17/50
23/23 - 0s - loss: 1544.3312
Epoch 18/50
23/23 - 0s - loss: 1537.1790
Epoch 19/50
23/23 - 0s - loss: 1529.3099
Epoch 20/50
23/23 - 0s - loss: 1520.5131
Epoch 21/50
23/23 - 0s - loss: 1510.6018
Epoch 22/50
23/23 - 0s - loss: 1499.7072
Epoch 23/50
23/23 - 0s - loss: 1487.5879
Epoch 24/50
23/23 - 0s - loss: 1474.3489
Epoch 25/50
23/23 - 0s -

Next we need to evaluate the model on the test data.

In [18]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 1s 1ms/step - loss: 630.2323


630.2322998046875

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [19]:
from sklearn.metrics import mean_squared_error

In [20]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

630.2322793411709 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [21]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 156.6608123779297
MSE 2: 161.0012664794922
MSE 3: 112.69192504882812
MSE 4: 101.87218475341797
MSE 5: 84.30003356933594
MSE 6: 71.43851470947266
MSE 7: 65.97096252441406
MSE 8: 49.29180145263672
MSE 9: 51.40507507324219
MSE 10: 47.8912467956543
MSE 11: 45.92076110839844
MSE 12: 44.337772369384766
MSE 13: 51.62886047363281
MSE 14: 48.76987838745117
MSE 15: 44.457618713378906
MSE 16: 40.68373107910156
MSE 17: 44.72843933105469
MSE 18: 42.95050048828125
MSE 19: 41.56831359863281
MSE 20: 42.80036544799805
MSE 21: 40.860939025878906
MSE 22: 40.256431579589844
MSE 23: 38.77655792236328
MSE 24: 42.62125015258789
MSE 25: 44.404823303222656
MSE 26: 42.81116485595703
MSE 27: 42.748355865478516
MSE 28: 39.30951690673828
MSE 29: 46.154632568359375
MSE 30: 43.02418518066406
MSE 31: 42.982322692871094
MSE 32: 40.95514678955078
MSE 33: 40.046875
MSE 34: 45.20136642456055
MSE 35: 42.29946517944336
MSE 36: 48.334712982177734
MSE 37: 43.001708984375
MSE 38: 46.45442581176758
MSE 39: 43.6117439270

So we got mean value 53.087 which is less than the previous value 76.812 and std is now 25.935 which was 32.791 previously.